irony-cdb-libclang.el

```diff
@@ -0,0 +1,89 @@
+;;; irony-cdb-libclang.el --- Compilation Database for irony using libclang
+
+;; Copyright (C) 2014  Guillaume Papin
+
+;; Author: Guillaume Papin <guillaume.papin@epitech.eu>
+;; Keywords: c, convenience, tools
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;;
+;; Compilation Database support for Irony using libclangs CXCompilationDatabase.
+;; Currently libclang only supports json compilation databases, but reading
+;; using libclang is much faster than using json.el
+
+;;; Code:
+
+(require 'irony-cdb)
+(require 'cl-lib)
+(require 'pp)
+
+;;;###autoload
+(defun irony-cdb-libclang (command &rest args)
+  (cl-case command
+    (get-compile-options (irony-cdb-libclang--get-compile-options))))
+
+(defvar-local irony-cdb-libclang--server-compile-options nil
+  "Compiler options got from irony-server.")
+
+(defun irony-cdb-libclang--request-handler (flags result)
+  (set result flags))
+
+;; TODO: sometimes there is garbage on the callback-queue, determine why
+(defun irony-cdb-libclang--get-compile-options ()
+  (irony--awhen (irony-cdb-json--locate-db)
+    (let ((db-location it))
+      ;; Try to obtain flags from server
+      (irony--aif (irony-cdb-libclang--server-exact-flags db-location)
+          it
+        ;; Couldn't get flags from server. Load db using irony-cdb-json guess
+        ;; the flags
+        (let* ((db (irony-cdb-json--load-db db-location))
+               (dir-cdb (irony-cdb-json--compute-directory-cdb db)))
+          (irony-cdb-json--guess-flags dir-cdb))))))
+
+(defun irony-cdb-libclang--server-exact-flags (db-file)
+  "Get compile options from server"
+  (let ((project-root (file-name-directory db-file))
+          (file buffer-file-truename))
```

Comment:
Careful, what if you have no file attached to this buffer?
`buffer-file-truename` returns the abbreviated filename, i.e: `~/`, I don't think that's what we want.

`file` is used for two things:

1. sending as argument to irony-server's get-compile-options, here we don't want to use `buffer-file-truename` but `buffer-file-name`. Tthis one doesn't exhibit the abbreviated-filename behavior, and the CXCompilationDatabase has the logic to handle truename if that is needed.
2. `irony-cdb-libclang--adjust-options-and-remove-compiler` that forwards it to `irony-cdb-json--adjust-compile-options`. Here we also want the original filename, `irony-cdb-json--adjust-compile-options` will should be responsible of calling truename if need be.

See `irony--get-buffer-path-for-server` logic too.